crates/modelardb_server/src/storage/uncompressed_data_manager.rs

```diff
@@ -93,61 +117,110 @@ impl UncompressedDataManager {
-            active_uncompressed_data_buffers: HashMap::new(),
-            finished_uncompressed_data_buffers: finished_data_buffers,
-            memory_pool,
-            compress_directly,
-            used_uncompressed_memory_metric: Metric::new(),
-            ingested_data_points_metric: Metric::new(),
-            used_disk_space_metric,
-        })
+        Ok(())
+    }
+
+    /// Read and process messages received from the [`super::StorageEngine`] to either ingest
+    /// uncompressed data, flush buffers, or stop.
+    pub(super) fn process_uncompressed_messages(
+        &self,
+        runtime: Arc<Runtime>,
+    ) -> Result<(), String> {
+        loop {
+            let message = self
+                .channels
+                .multivariate_data_receiver
+                .recv()
+                .map_err(|error| error.to_string())?;
+
+            match message {
+                Message::Data(uncompressed_data_multivariate) => {
+                    runtime
+                        .block_on(self.insert_data_points(uncompressed_data_multivariate))
+                        .map_err(|error| error.to_string())?;
+                }
+                Message::Flush => {
+                    self.flush().unwrap();
+                    self.channels
+                        .univariate_data_sender
+                        .send(Message::Flush)
+                        .map_err(|error| error.to_string())?;
+                }
+                Message::Stop => {
+                    self.flush().unwrap();
+                    self.channels
+                        .univariate_data_sender
+                        .send(Message::Stop)
+                        .map_err(|error| error.to_string())?;
+                    break;
+                }
+            }
+        }
+
+        Ok(())
     }
 
-    // TODO: When the compression component is changed, this function should return Ok or Err.
-    /// Split `data_points` and insert them into the in-memory buffer. The multivariate time series
-    /// in `data_points` are first split into multiple univariate time series based on
-    /// `model_table`. These univariate time series are then inserted into the storage engine.
-    /// Return a list of buffers that were compressed as a result of inserting the data points with
-    /// their corresponding univariate id if the data was successfully inserted, otherwise return
-    /// [`Err`].
-    pub(super) async fn insert_data_points(
-        &mut self,
-        metadata_manager: &MetadataManager,
-        model_table_metadata: &ModelTableMetadata,
-        data_points: &RecordBatch,
-    ) -> Result<Vec<(u64, RecordBatch)>, String> {
+    /// Insert `uncompressed_data_multivariate` into the in-memory buffer. If the data points are
+    /// from a multivariate time series, they are first split into multiple univariate time series.
+    /// These univariate time series are then inserted into the storage engine. Return [`Err`] if
```

Comment:
Fixed. I agree that our error handling could be significantly improved, e.g., by using our own error types instead of built-in and strings, and there is a task about looking into what the current recommendations for error handling is in Rust and then implement them.